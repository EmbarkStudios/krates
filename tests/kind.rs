mod util;

use util::build;

#[test]
fn all_the_things() {
    let grafs = build("all-features.json", krates::Builder::new()).unwrap();

    assert!(grafs.filtered.is_empty());
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        // This shouldn't actually affect anything, as dev dependencies
        // for non-workspace crates are already not resolved
        assert!(grafs.filtered.is_empty());
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        // This will be equivalent to to filtering workspace dev crates
        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_normal() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build_and_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn only_b() {
    let mut kb = krates::Builder::new();
    kb.include_workspace_crates(["/home/jake/code/krates/tests/ws/b/Cargo.toml"]);

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_after_build() {
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Dev);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }

    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Build);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }
}
