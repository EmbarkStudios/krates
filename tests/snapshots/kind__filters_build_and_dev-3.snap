---
source: tests/kind.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 path+file:///krates/tests/ws/a" ]
    1 [ label = "crate b 0.1.0 path+file:///krates/tests/ws/b" ]
    2 [ label = "crate bitflags 1.3.2" ]
    3 [ label = "crate bumpalo 3.11.0" ]
    4 [ label = "crate c 0.1.0 path+file:///krates/tests/ws/c" ]
    5 [ label = "crate cfg-if 0.1.10" ]
    6 [ label = "crate cfg-if 1.0.0" ]
    7 [ label = "crate coreaudio-rs 0.9.1" ]
    8 [ label = "crate coreaudio-sys 0.2.10" ]
    9 [ label = "crate js-sys 0.3.60" ]
    10 [ label = "crate lazy_static 1.4.0" ]
    11 [ label = "crate leftpad 0.2.0" ]
    12 [ label = "crate libc 0.2.133" ]
    13 [ label = "crate log 0.4.17" ]
    14 [ label = "crate nix 0.16.1" ]
    15 [ label = "crate once_cell 1.15.0" ]
    16 [ label = "crate proc-macro2 1.0.44" ]
    17 [ label = "crate quote 1.0.21" ]
    18 [ label = "crate spin 0.5.2" ]
    19 [ label = "crate syn 1.0.101" ]
    20 [ label = "crate unicode-ident 1.0.4" ]
    21 [ label = "crate void 1.0.2" ]
    22 [ label = "crate wasm-bindgen 0.2.83" ]
    23 [ label = "crate wasm-bindgen-backend 0.2.83" ]
    24 [ label = "crate wasm-bindgen-futures 0.4.33" ]
    25 [ label = "crate wasm-bindgen-macro 0.2.83" ]
    26 [ label = "crate wasm-bindgen-macro-support 0.2.83" ]
    27 [ label = "crate wasm-bindgen-shared 0.2.83" ]
    28 [ label = "crate web-sys 0.3.60" ]
    29 [ label = "crate winapi 0.2.8" ]
    30 [ label = "feature default" ]
    31 [ label = "feature default" ]
    32 [ label = "feature Crypto" ]
    33 [ label = "feature Window" ]
    34 [ label = "feature default" ]
    35 [ label = "feature audio_toolbox" ]
    36 [ label = "feature audio_unit" ]
    37 [ label = "feature core_audio" ]
    38 [ label = "feature core_midi" ]
    39 [ label = "feature open_al" ]
    40 [ label = "feature default" ]
    41 [ label = "feature extra_traits" ]
    42 [ label = "feature default" ]
    43 [ label = "feature default" ]
    44 [ label = "feature proc-macro" ]
    45 [ label = "feature proc-macro" ]
    46 [ label = "feature spans" ]
    47 [ label = "feature default" ]
    48 [ label = "feature default" ]
    49 [ label = "feature default" ]
    50 [ label = "feature default" ]
    51 [ label = "feature full" ]
    52 [ label = "feature default" ]
    53 [ label = "feature MessageEvent" ]
    54 [ label = "feature Worker" ]
    55 [ label = "feature spans" ]
    56 [ label = "feature visit" ]
    57 [ label = "feature spans" ]
    58 [ label = "feature leftpad" ]
    59 [ label = "feature leftier-strings" ]
    60 [ label = "feature lazy_static" ]
    61 [ label = "feature open_al" ]
    62 [ label = "feature audio_toolbox" ]
    63 [ label = "feature audio_unit" ]
    64 [ label = "feature core_audio" ]
    65 [ label = "feature core_midi" ]
    66 [ label = "feature std" ]
    67 [ label = "feature std" ]
    68 [ label = "feature alloc" ]
    69 [ label = "feature race" ]
    70 [ label = "feature quote" ]
    71 [ label = "feature proc-macro" ]
    72 [ label = "feature printing" ]
    73 [ label = "feature parsing" ]
    74 [ label = "feature derive" ]
    75 [ label = "feature clone-impls" ]
    76 [ label = "feature std" ]
    77 [ label = "feature std" ]
    78 [ label = "feature spans" ]
    79 [ label = "feature EventTarget" ]
    80 [ label = "feature Event" ]
    0 -> 1 [ label = "" ]
    1 -> 30 [ label = "" ]
    1 -> 24 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    1 -> 24 [ label = " 'cfg(all(target_vendor = \"xboxone\"))'" ]
    4 -> 31 [ label = " 'x86_64-apple-darwin'" ]
    4 -> 10 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    4 -> 11 [ label = "" ]
    4 -> 12 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    4 -> 14 [ label = " 'x86_64-unknown-linux-gnu'" ]
    4 -> 18 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    4 -> 32 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    4 -> 33 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    4 -> 29 [ label = " 'cfg(target_os = \"windows\")'" ]
    7 -> 34 [ label = "" ]
    7 -> 8 [ label = "" ]
    7 -> 35 [ label = "" ]
    7 -> 36 [ label = "" ]
    7 -> 37 [ label = "" ]
    7 -> 38 [ label = "" ]
    7 -> 39 [ label = "" ]
    9 -> 40 [ label = "" ]
    13 -> 6 [ label = "" ]
    14 -> 34 [ label = "" ]
    14 -> 5 [ label = "" ]
    14 -> 41 [ label = "" ]
    14 -> 42 [ label = "" ]
    14 -> 43 [ label = "" ]
    16 -> 20 [ label = "" ]
    17 -> 16 [ label = "" ]
    17 -> 44 [ label = "" ]
    19 -> 16 [ label = "" ]
    19 -> 44 [ label = "" ]
    19 -> 17 [ label = "" ]
    19 -> 45 [ label = "" ]
    19 -> 20 [ label = "" ]
    22 -> 6 [ label = "" ]
    22 -> 25 [ label = "" ]
    22 -> 46 [ label = "" ]
    23 -> 47 [ label = "" ]
    23 -> 13 [ label = "" ]
    23 -> 48 [ label = "" ]
    23 -> 49 [ label = "" ]
    23 -> 50 [ label = "" ]
    23 -> 51 [ label = "" ]
    23 -> 52 [ label = "" ]
    23 -> 27 [ label = "" ]
    24 -> 6 [ label = "" ]
    24 -> 9 [ label = "" ]
    24 -> 40 [ label = "" ]
    24 -> 53 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    24 -> 54 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    25 -> 50 [ label = "" ]
    25 -> 26 [ label = "" ]
    25 -> 55 [ label = "" ]
    26 -> 49 [ label = "" ]
    26 -> 50 [ label = "" ]
    26 -> 56 [ label = "" ]
    26 -> 51 [ label = "" ]
    26 -> 52 [ label = "" ]
    26 -> 23 [ label = "" ]
    26 -> 57 [ label = "" ]
    26 -> 27 [ label = "" ]
    28 -> 9 [ label = "" ]
    28 -> 40 [ label = "" ]
    34 -> 2 [ label = "" ]
    47 -> 3 [ label = "" ]
    58 -> 4 [ label = "" ]
    58 -> 11 [ label = "" ]
    59 -> 4 [ label = "" ]
    59 -> 58 [ label = "" ]
    60 -> 4 [ label = "" ]
    60 -> 10 [ label = "" ]
    30 -> 4 [ label = "" ]
    30 -> 58 [ label = "" ]
    61 -> 7 [ label = "" ]
    61 -> 39 [ label = "" ]
    31 -> 7 [ label = "" ]
    31 -> 62 [ label = "" ]
    31 -> 63 [ label = "" ]
    31 -> 64 [ label = "" ]
    31 -> 61 [ label = "" ]
    31 -> 65 [ label = "" ]
    65 -> 7 [ label = "" ]
    65 -> 38 [ label = "" ]
    64 -> 7 [ label = "" ]
    64 -> 37 [ label = "" ]
    63 -> 7 [ label = "" ]
    63 -> 36 [ label = "" ]
    62 -> 7 [ label = "" ]
    62 -> 35 [ label = "" ]
    39 -> 8 [ label = "" ]
    38 -> 8 [ label = "" ]
    37 -> 8 [ label = "" ]
    36 -> 8 [ label = "" ]
    35 -> 8 [ label = "" ]
    66 -> 12 [ label = "" ]
    41 -> 12 [ label = "" ]
    42 -> 12 [ label = "" ]
    42 -> 66 [ label = "" ]
    67 -> 15 [ label = "" ]
    67 -> 68 [ label = "" ]
    69 -> 15 [ label = "" ]
    48 -> 15 [ label = "" ]
    48 -> 67 [ label = "" ]
    68 -> 15 [ label = "" ]
    68 -> 69 [ label = "" ]
    44 -> 16 [ label = "" ]
    49 -> 16 [ label = "" ]
    49 -> 44 [ label = "" ]
    45 -> 17 [ label = "" ]
    45 -> 44 [ label = "" ]
    50 -> 17 [ label = "" ]
    50 -> 45 [ label = "" ]
    56 -> 19 [ label = "" ]
    70 -> 19 [ label = "" ]
    70 -> 17 [ label = "" ]
    71 -> 19 [ label = "" ]
    71 -> 44 [ label = "" ]
    71 -> 45 [ label = "" ]
    72 -> 19 [ label = "" ]
    72 -> 70 [ label = "" ]
    73 -> 19 [ label = "" ]
    51 -> 19 [ label = "" ]
    74 -> 19 [ label = "" ]
    52 -> 19 [ label = "" ]
    52 -> 74 [ label = "" ]
    52 -> 73 [ label = "" ]
    52 -> 72 [ label = "" ]
    52 -> 75 [ label = "" ]
    52 -> 71 [ label = "" ]
    75 -> 19 [ label = "" ]
    76 -> 21 [ label = "" ]
    43 -> 21 [ label = "" ]
    43 -> 76 [ label = "" ]
    77 -> 22 [ label = "" ]
    78 -> 22 [ label = "" ]
    78 -> 46 [ label = "" ]
    40 -> 22 [ label = "" ]
    40 -> 78 [ label = "" ]
    40 -> 77 [ label = "" ]
    57 -> 23 [ label = "" ]
    46 -> 25 [ label = "" ]
    46 -> 55 [ label = "" ]
    55 -> 26 [ label = "" ]
    55 -> 57 [ label = "" ]
    54 -> 28 [ label = "" ]
    54 -> 79 [ label = "" ]
    33 -> 28 [ label = "" ]
    33 -> 79 [ label = "" ]
    53 -> 28 [ label = "" ]
    53 -> 80 [ label = "" ]
    79 -> 28 [ label = "" ]
    80 -> 28 [ label = "" ]
    32 -> 28 [ label = "" ]
}

